std: fix stdout-before-main

joboet opened the pull request:
Fixes #130210.

Since #124881, `ReentrantLock` uses `ThreadId` to identify threads. This has the unfortunate consequence of breaking uses of `Stdout` before main: Locking the `ReentrantLock` that synchronizes the output will initialize the thread ID before the handle for the main thread is set in `rt::init`. But since that would overwrite the current thread ID, `thread::set_current` triggers an abort.

This PR fixes the problem by using the already initialized thread ID for constructing the main thread handle and allowing `set_current` calls that do not change the thread's ID.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

tgross35 commented on tests/ui/runtime/stdout-before-main.run.stdout:
```diff
@@ -0,0 +1 @@
+Hello from before main!
```
Could the call in `main` be changed to `println` so this file doesn't wind up without a newline?

joboet replied:
Sure!

tgross35 commented:
One nit if it makes sense, r=me either way

joboet committed:
std: fix stdout-before-main

Fixes #130210.

Since #124881, `ReentrantLock` uses `ThreadId` to identify threads. This has the unfortunate consequence of breaking uses of `Stdout` before main: Locking the `ReentrantLock` that synchronizes the output will initialize the thread ID before the handle for the main thread is set in `rt::init`. But since that would overwrite the current thread ID, `thread::set_current` triggers an abort.

This PR fixes the problem by using the already initialized thread ID for constructing the main thread handle and allowing `set_current` calls that do not change the thread's ID.

joboet commented:
@bors r=@tgross35

bors commented:
:pushpin: Commit 9f91c5099fb7261cf8c85cc638d2692a317060ec has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 9f91c5099fb7261cf8c85cc638d2692a317060ec -->
<!-- homu: {"type":"Approved","sha":"9f91c5099fb7261cf8c85cc638d2692a317060ec","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
